src/otaclient/create_standby/delta_gen.py

```diff
@@ -0,0 +1,743 @@
+from ota_metadata.legacy2.rs_table import ResourceTableORMPool
+from otaclient._status_monitor import StatusReport, UpdateProgressReport
+from otaclient.configs.cfg import cfg
+from otaclient.create_standby.utils import TopDownCommonShortestPath
+from otaclient_common import EMPTY_FILE_SHA256, EMPTY_FILE_SHA256_BYTE, replace_root
+from otaclient_common._io import _gen_tmp_fname
+from otaclient_common._typing import StrOrPath
+
+logger = logging.getLogger(__name__)
+
+T = TypeVar("T")
+
+CANONICAL_ROOT = cfg.CANONICAL_ROOT
+CANONICAL_ROOT_P = Path(CANONICAL_ROOT)
+
+DB_CONN_NUMS = 3
+
+PROCESS_FILES_REPORT_INTERVAL = cfg.PROCESS_FILES_REPORT_INTERVAL
+
+
+class UpdateStandbySlotFailed(Exception): ...
+
+
+class DeltaGenParams(TypedDict):
+    ota_metadata: OTAMetadata
+    delta_src: Path
+    copy_dst: Path
+    status_report_queue: Queue[StatusReport]
+    session_id: str
+
+
+class _DeltaGeneratorBase:
+    """
+    NOTE: the instance of this class cannot be re-used after delta is generated.
+    """
+
+    CLEANUP_ENTRY = {Path("/lost+found"), Path("/tmp"), Path("/run")}
+    """Entries we need to cleanup everytime, include /lost+found, /tmp and /run."""
+
+    OTA_WORK_PATHS = {Path(cfg.OTA_TMP_STORE), Path(cfg.OTA_TMP_META_STORE)}
+    """Folders for otaclient working on standby slot."""
+
+    def __init__(
+        self,
+        *,
+        ota_metadata: OTAMetadata,
+        delta_src: Path,
+        copy_dst: Path,
+        status_report_queue: Queue[StatusReport],
+        session_id: str,
+    ) -> None:
+        self._status_report_queue = status_report_queue
+        self.session_id = session_id
+        self._delta_src_mount_point = delta_src
+        self._copy_dst = copy_dst
+        self._ota_metadata = ota_metadata
+
+        self._ft_reg_orm = FileTableRegularORMPool(
+            con_factory=ota_metadata.connect_fstable, number_of_cons=DB_CONN_NUMS
+        )
+        self._ft_dir_orm = FileTableDirORM(ota_metadata.connect_fstable())
+        self._rst_orm_pool = ResourceTableORMPool(
+            con_factory=ota_metadata.connect_rstable, number_of_cons=DB_CONN_NUMS
+        )
+
+        self._que = Queue()
+        self._max_pending_tasks = threading.Semaphore(
+            cfg.MAX_CONCURRENT_PROCESS_FILE_TASKS
+        )
+        self._dirs_to_remove = dtr = TopDownCommonShortestPath()
+        for _p in self.CLEANUP_ENTRY:
+            dtr.add_path(Path(_p))
+
+        # put the empty file into copy_dst, remember to commit a record for preparing the empty file
+        (copy_dst / EMPTY_FILE_SHA256).touch()
+        self._rst_orm_pool.orm_delete_entries(digest=EMPTY_FILE_SHA256_BYTE)
+        self._status_report_queue.put_nowait(
+            StatusReport(
+                payload=UpdateProgressReport(
+                    operation=UpdateProgressReport.Type.PREPARE_LOCAL_COPY,
+                    processed_file_num=1,
+                ),
+                session_id=self.session_id,
+            )
+        )
```

Comment:
Handle empty file, prepare the empty file in advance and remove it from the resource database.